Empty source tables are not deleted from global configuration file

- [x] I have searched the issue tracker and believe that this is not a duplicate.

**Make sure you run commands with `-v` flag before pasting the output.**

## Steps to reproduce

<!--Describe the minimized example of how to reproduce the bug-->

1. Add a source URL to the global configuration:
   
   ```shell
   pdm config pypi.custom.url https://test.pypi.org/simple/
   ```

2. Remove the source URL again:

   ```shell
   pdm config pypi.custom.url --delete
   ```

3. The `config.toml` file now contains an empty table `[pypi.custom]`:

   ```toml
   [pypi]
   [pypi.custom]
   ```

4. Execute a PDM command in any of your PDM managed projects, e.g. `pdm lock`
5. See the error `[AssertionError]: Source URL must not be empty`:

   ```shell
   PS F:\my-project> pdm lock -v
   Traceback (most recent call last):
      File "<frozen runpy>", line 198, in _run_module_as_main
      File "<frozen runpy>", line 88, in _run_code
      File "c:\users\p.rogalski\.local\bin\pdm.exe\__main__.py", line 7, in <module>
      File "C:\Users\p.rogalski\.local\pipx\venvs\pdm\Lib\site-packages\pdm\core.py", line 289, in main
      return Core().main(args)
            ^^^^^^^^^^^^^^^^^
      File "C:\Users\p.rogalski\.local\pipx\venvs\pdm\Lib\site-packages\pdm\core.py", line 207, in main
      raise cast(Exception, err).with_traceback(traceback) from None
      File "C:\Users\p.rogalski\.local\pipx\venvs\pdm\Lib\site-packages\pdm\core.py", line 202, in main
      self.handle(project, options)
      File "C:\Users\p.rogalski\.local\pipx\venvs\pdm\Lib\site-packages\pdm\core.py", line 157, in handle
      command.handle(project, options)
      File "C:\Users\p.rogalski\.local\pipx\venvs\pdm\Lib\site-packages\pdm\cli\commands\lock.py", line 61, in handle
      actions.do_lock(
      File "C:\Users\p.rogalski\.local\pipx\venvs\pdm\Lib\site-packages\pdm\cli\actions.py", line 70, in do_lock
      provider = project.get_provider(strategy, tracked_names, ignore_compatibility=cross_platform)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "C:\Users\p.rogalski\.local\pipx\venvs\pdm\Lib\site-packages\pdm\project\core.py", line 426, in get_provider
      repository = self.get_repository(ignore_compatibility=ignore_compatibility)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "C:\Users\p.rogalski\.local\pipx\venvs\pdm\Lib\site-packages\pdm\project\core.py", line 393, in get_repository
      sources = self.sources or []
                  ^^^^^^^^^^^^
      File "C:\Users\p.rogalski\.local\pipx\venvs\pdm\Lib\site-packages\pdm\project\core.py", line 383, in sources
      assert source.url, "Source URL must not be empty"
   AssertionError: Source URL must not be empty
   ```

## Actual behavior

<!--A clear and concise description the result of the above steps-->

PDM errors out on almost any comand, until I manually delete the table `[pypi.custom]` from `C:\Users\p.rogalski\AppData\Local\pdm\pdm\config.toml`.

## Expected behavior

<!--A clear and concise description of what you expected to happen.-->

I should be able to continue working with PDM.

## Environment Information

```bash
# Paste the output of `pdm info && pdm info --env` below:
PS F:\my-project> pdm info && pdm info --env
PDM version:
  2.7.4
Python Interpreter:
  F:\my-project\.venv\Scripts\python.exe (3.11-32)
Project Root:
  F:/my-project
Local Packages:
  
{
  "implementation_name": "cpython",
  "implementation_version": "3.11.2",
  "os_name": "nt",
  "platform_machine": "AMD64",
  "platform_release": "10",
  "platform_system": "Windows",
  "platform_version": "10.0.22621",
  "python_full_version": "3.11.2",
  "platform_python_implementation": "CPython",
  "python_version": "3.11",
  "sys_platform": "win32"
}
```

I had to manually delete the table `[pypi.custom]` from `C:\Users\p.rogalski\AppData\Local\pdm\pdm\config.toml` for these commands to work.

--- a/news/2034.bugfix.md
+++ b/news/2034.bugfix.md
@@ -1,0 +1,1 @@
+Fix a bug that empty source tables in configuration files causes errors when running pdm commands.

--- a/src/pdm/project/config.py
+++ b/src/pdm/project/config.py
@@ -260,7 +260,7 @@ class Config(MutableMapping[str, str]):
 
     def iter_sources(self) -> Iterator[RepositoryConfig]:
         for name, data in self._data.items():
-            if name.startswith(f"{SOURCE}.") and name not in self._config_map:
+            if name.startswith(f"{SOURCE}.") and name not in self._config_map and data:
                 yield RepositoryConfig(**data, name=name[len(SOURCE) + 1 :], config_prefix=SOURCE)
 
     def _save_config(self) -> None:

--- a/src/pdm/project/core.py
+++ b/src/pdm/project/core.py
@@ -380,7 +380,7 @@ class Project:
             merge_sources(self.project_config.iter_sources())
             merge_sources(self.global_config.iter_sources())
         for source in result.values():
-            assert source.url, "Source URL must not be empty"
+            assert source.url, f"Source URL must not be empty for {source.name}"
             source.url = expand_env_vars_in_auth(source.url)
         return list(result.values())
 

